### PR TITLE
U4-10669 7.8. Show 301 url redirects on info tab too

### PR DESCRIPTION
**Pull request by marcemarc**
Following on from this blog post: http://tooorangey.co.uk/posts/301-redirection-in-umbraco-it-s-a-rum-do/ found this issue http://issues.umbraco.org/issue/U4-10669 raised by @JeffreyPerplex  and @hartvig seemed to suggest 'I agree it should be on the properties tab, though and would be happy to accept a PR that placed it there'

This is then that PR.

But it's not clear from the issue, if we're all thumbs up for adding the ability for editors to manually add redirects here, or on the redirect url management dashboard...yet

... so this PR just covers the bit that is a no brainer - you should be able to find redirects based on the page they redirect to!

Some bits I wasn't sure about:

1) where in the info tab is the Redirect Url Management box best placed?
2) should the box show if there are no redirects?
3) The markup... apologies to @madsrasmussen I've used the markup from the Redirect Url Management dashboard here.... the remove buttons dominate the whole page! 
4) I've branched from dev-7.8 as that's what had the new info tab layout in...
5) What should happen if the Redirect Url Management tracker is disabled, eg via the dashboard or because another 301 package is installed? should this box show on the info tab in those circumstances?

Anyway hopefully this is of some use for fangling this functionality into order!

I'll let the 'manually adding a redirect stuff' sit in a separate branch so as not to confuse this bit.

### CODE REVIEW

**Comment by dawoe** [comment]
Hi @marcemarc 

Thx for this PR. I'm testing this one at the moment.

Will give some feedback soon.

Dave [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-cpr-team/)

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html [hunk 1]
```diff
@@ -19,6 +19,34 @@
                 </ul>
             </umb-box-content>
         </umb-box>
+        <umb-box data-element="node-info-redirects">
+            <umb-box-header title-key="redirectUrls_redirectUrlManagement"></umb-box-header>
+            <umb-box-content class="block-form">
+                <div style="position: relative;">
+                    <div ng-if="loadingRedirectUrls" style="background: rgba(255, 255, 255, 0.8); position: absolute; top: 0; left: 0; right: 0; bottom: 0;"></div>
+                    <umb-load-indicator ng-if="loadingRedirectUrls"></umb-load-indicator>
+                    <!--<ul class="nav nav-stacked" style="margin-bottom: 0;" ng-show="hasRedirects">
```
Comment (dawoe) [inline src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: Can this commented code be removed ?

**Comment by marcemarc** [comment]
Hi @dawoe updated!

Removed the ability to remove the redirects, and updated html!

Added check on whether Redirect Url Management dashboard is enabled or not...

... as always... 

not sure if it's done in the best way possible!

regards

Marc

File: src/Umbraco.Web.UI.Client/src/common/directives/components/content/umbcontentnodeinfo.directive.js [hunk 2]
```diff
@@ -95,23 +96,43 @@
...
                         scope.loadingAuditTrail = false;
                     });
 
             }
+            function loadRedirectUrls() {
+                scope.loadingRedirectUrls = true;
+                //check if Redirect Url Management is enabled
+                redirectUrlsResource.getEnableState().then(function (response) {
+                    scope.urlTrackerDisabled = response.enabled !== true;
+                    scope.userIsAdmin = response.userIsAdmin;
```
Comment (dawoe) [inline src/Umbraco.Web.UI.Client/src/common/directives/components/content/umbcontentnodeinfo.directive.js]: Hi @marcemarc 

Where is this used ? Can't seem to find any uses

Dave

Comment (marcemarc) [reply src/Umbraco.Web.UI.Client/src/common/directives/components/content/umbcontentnodeinfo.directive.js]: It's in the view:

 <div ng-show="!hasRedirects && !loadingRedirectUrls && !urlTrackerDisabled"><localize key="redirectUrls_noRedirects"></localize></div>
                    <div ng-show="urlTrackerDisabled">Redirect Url Management is disabled</div>

Essentially I'm thinking if the redirect url management dashboard is disabled, then we shouldn't show the redirect urls here?
interestingly the getEnableState endpoint returns whether the user is an admin, but I'm doing anything with it here, it just got me thinking from your comments about not letting just anyone remove redirects, anyway as we've removed that too, shall I remove the userIsAdmin setting too for now?

**Comment by marcemarc** [comment]
@dawoe removed the userIsAdmin bit, in case that was what you were highlighting?

**Comment by dawoe** [comment]
Yep..was talking about that one. Will try to retest this one today

File: src/Umbraco.Web/Editors/RedirectUrlManagementController.cs [hunk 3]
```diff
@@ -57,7 +58,29 @@ public RedirectUrlSearchResult SearchRedirectUrls(string searchTerm, int page =
...
+        /// </summary>
+        /// <param name="contentUdi">Udi of content item to retrieve RedirectUrls for</param>
+        /// <returns></returns>
+        [HttpGet]
+        public RedirectUrlSearchResult RedirectUrlsForContentItem(string contentUdi)
+        {
+            var redirectsResult = new RedirectUrlSearchResult();
+            GuidUdi guidIdi;
+            var contentKey = GuidUdi.TryParse(contentUdi, out guidIdi) ? guidIdi.Guid : default(Guid);
+            //hmm what to do if this doesn't parse as a Guid?
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web/Editors/RedirectUrlManagementController.cs]: I would just return an empty result if anything fails:

```
        public RedirectUrlSearchResult RedirectUrlsForContentItem(string contentUdi)
        {
            var redirectsResult = new RedirectUrlSearchResult();
            if (GuidUdi.TryParse(contentUdi, out var guidIdi))
            {
                var redirectUrlService = Services.RedirectUrlService;
                var redirects = redirectUrlService.GetContentRedirectUrls(guidIdi.Guid);
                redirectsResult.SearchResults = Mapper.Map<IEnumerable<ContentRedirectUrl>>(redirects).ToArray();
                // now map the Content/published - don't need to do this? - not displaying the target
                // Url on the info tab, no need to retrieve in the same was as on the dashboard
                // not doing paging 'yet'
                redirectsResult.TotalCount = redirects.Count();
                redirectsResult.CurrentPage = 1;
                redirectsResult.PageCount = 1;
            }

            return redirectsResult;
        }
```

File: src/Umbraco.Web.UI/umbraco/config/lang/zh.xml [hunk 4]
```diff
@@ -1606,6 +1606,7 @@
     <key alias="enableUrlTracker">启用 url 跟踪程序</key>
     <key alias="originalUrl">原始网址</key>
     <key alias="redirectedTo">已重定向至</key>
+    <key alias="redirectUrlManagement">Redirect Url Management</key>
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.UI/umbraco/config/lang/zh.xml]: Can you revert the changes to all the language files please? New keys only need to be in `en.xml` and `en_us.xml`, the latter is used for fallbacks if a key is not available in the user's language.

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html [hunk 5]
```diff
@@ -19,6 +19,22 @@
...
+                <div style="position: relative;">
+                    <div ng-if="loadingRedirectUrls" style="background: rgba(255, 255, 255, 0.8); position: absolute; top: 0; left: 0; right: 0; bottom: 0;"></div>
+                    <umb-load-indicator ng-if="loadingRedirectUrls"></umb-load-indicator>
+                    <ul class="nav nav-stacked" style="margin-bottom: 0;" ng-show="hasRedirects">
+                        <li ng-repeat="redirectUrl in redirectUrls">
+                            <a href="{{redirectUrl.originalUrl}}"><i ng-class="value.icon" class="icon-out"></i> {{redirectUrl.originalUrl}}</a>
+                        </li>
+                    </ul>
+                    <div ng-show="!hasRedirects && !loadingRedirectUrls && !urlTrackerDisabled"><localize key="redirectUrls_noRedirects"></localize></div>
+                    <div ng-show="urlTrackerDisabled">Redirect Url Management is disabled</div>
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: Instead of showing: we can't show you information, I think it would be better to not show this panel altogether.

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html [hunk 6]
```diff
@@ -19,6 +19,22 @@
...
         </umb-box>
+        <umb-box data-element="node-info-redirects">
+            <umb-box-header title-key="redirectUrls_redirectUrlManagement"></umb-box-header>
+            <umb-box-content class="block-form">
+                <div style="position: relative;">
+                    <div ng-if="loadingRedirectUrls" style="background: rgba(255, 255, 255, 0.8); position: absolute; top: 0; left: 0; right: 0; bottom: 0;"></div>
+                    <umb-load-indicator ng-if="loadingRedirectUrls"></umb-load-indicator>
+                    <ul class="nav nav-stacked" style="margin-bottom: 0;" ng-show="hasRedirects">
+                        <li ng-repeat="redirectUrl in redirectUrls">
+                            <a href="{{redirectUrl.originalUrl}}"><i ng-class="value.icon" class="icon-out"></i> {{redirectUrl.originalUrl}}</a>
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: Be good to add `target="_blank"`  here.

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html [hunk 7]
```diff
@@ -19,6 +19,22 @@
                 </ul>
             </umb-box-content>
         </umb-box>
+        <umb-box data-element="node-info-redirects">
+            <umb-box-header title-key="redirectUrls_redirectUrlManagement"></umb-box-header>
+            <umb-box-content class="block-form">
+                <div style="position: relative;">
+                    <div ng-if="loadingRedirectUrls" style="background: rgba(255, 255, 255, 0.8); position: absolute; top: 0; left: 0; right: 0; bottom: 0;"></div>
+                    <umb-load-indicator ng-if="loadingRedirectUrls"></umb-load-indicator>
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: Once loaded and there are redirects, it would be good to show a one-liner like: 

> The following URLs are pointing to the current page:

It makes it more clear what this is, the tiny arrow pointing inwards is pretty easy to overlook. 😉

Comment (marcemarc) [reply src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: I have added this, ironically I had it in originally, but I always worry when I'm adding text explaining what something does, because it sort of means it isn't obvious from the UX, and it's more text to translate... perhaps it should be an item on the backoffice tour?

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html (hunk 6 above)
Comment (marcemarc) [reply src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: added

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html (hunk 5 above)
Comment (marcemarc) [reply src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html]: Because it's only clear after the api request has pulled back the redirects, whether there is anything to show, i was getting an annoying wobble when switching to that tab, eg you'd see the redirects panel for a second, then it would either disappear, or appear, with a flicker, I've added ng-cloak and an inline style of display:none, and I think the flicker has gone! - but obviously this probably isn't the 'bestest' front end way of achieve this... but I would take advice on how to do it super neatly?

File: src/Umbraco.Web.UI/umbraco/config/lang/zh.xml (hunk 4 above)
Comment (marcemarc) [reply src/Umbraco.Web.UI/umbraco/config/lang/zh.xml]: done, I always thought was being helpful adding the extra keys in!!

File: src/Umbraco.Web/Editors/RedirectUrlManagementController.cs (hunk 3 above)
Comment (marcemarc) [reply src/Umbraco.Web/Editors/RedirectUrlManagementController.cs]: cool, have updated, thanks!

**Comment by marcemarc** [comment]
@nul800sebastiaan hopefully have pushed and updated as per your comments

**Comment by nul800sebastiaan** [comment]
Apologies for the wait @marcemarc but it all looks very great now.. it's going in! 🎉 👍 ⭐️

File: src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html [hunk 8]
```diff
@@ -19,14 +19,12 @@
                 </ul>
             </umb-box-content>
         </umb-box>
-        <umb-box data-element="node-info-redirects">
+        <umb-box data-element="node-info-redirects" style="display:none;" ng-cloak ng-show="!urlTrackerDisabled && hasRedirects">
```
Comment (leekelleher) [inline src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html:22]: @marcemarc I'm (finally) using Redirects in vain, (my first 'real' project on v8) but I couldn't get the redirects panel to show on the Info app. I've pinpointed it down to the combination of `ng-show` and inline `display:none` - doesn't work together.

I swapped out the `style="display:none;" ng-cloak ng-show` with an `ng-if`, which works in my instance.

Before I start preparing a PR for this, I'm curious, did this work for you? (or anyone else?)

Comment (marcemarc) [reply src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html:22]: Hi @leekelleher this was added to V7 and worked at the time... and I've just been reminded to reply here, as I've just seen it on a V7 site we are currently working on! I'm guessing something got changed in V8 or in the Angular version change... I don't profess to understand why or how it did work... but it did work and still does! on V7! - but it's kinda hidden functionality so I doubt it would have been on anyone's radar to test.... I sort of think though it shouldn't be doing this show and hide boxy thing that only appears if redirects are present, and if I was creating it now, I'd perhaps implement the 'OtherUrls' function of a custom UrlProvider, to make the redirect Urls appear in the same Urls box as the main Url... but maybe it is better to separate it I don't know!

Comment (leekelleher) [reply src/Umbraco.Web.UI.Client/src/views/components/content/umb-content-node-info.html:22]: Follow up on this (if anyone sees this in future), this has been resolved for v8, in PR #8713.
